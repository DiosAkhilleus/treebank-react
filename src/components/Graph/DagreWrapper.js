import React, { useRef, useEffect } from 'react';
import { arrayOf, object, func } from 'prop-types';
import dagreD3 from 'dagre-d3';
import * as d3 from 'd3';

import styles from './Graph.module.scss';

const config = {
  rankdir: 'TB',
};

const drawGraph = (svg, g, { nodes, links, onClick }) => {
  const graph = new dagreD3.graphlib.Graph().setGraph(config);

  nodes.forEach(({
    id, label, config: { labelType, labelStyle, class: className },
  }) => {
    graph.setNode(id, {
      label, labelType, labelStyle, class: className,
    });
  });

  links.forEach(({
    source, target, label, config: { arrowheadStyle },
  }) => {
    graph.setEdge(source, target, { label, arrowheadStyle });
  });

  // eslint-disable-next-line new-cap
  const renderer = new dagreD3.render();
  const selectedSvg = d3.select(svg.current);
  const selectedG = d3.select(g.current);
  const zoom = d3.zoom().on('zoom', (event) => selectedG.attr('transform', event.transform));

  selectedSvg.call(zoom);

  renderer(selectedG, graph);

  selectedSvg.selectAll('g.node').on('click', ({ target }) => {
    // eslint-disable-next-line no-underscore-dangle
    onClick(target.__data__);
  });
};

const DagreWrapper = ({ nodes, links, onClick }) => {
  const svg = useRef(null);
  const g = useRef(null);

  useEffect(() => drawGraph(svg, g, { nodes, links, onClick }));

  return (
    <svg ref={svg} className={styles.graph}>
      <g ref={g} />
    </svg>
  );
};

DagreWrapper.propTypes = {
  nodes: arrayOf(object).isRequired,
  links: arrayOf(object).isRequired,
  onClick: func.isRequired,
};

export default DagreWrapper;
