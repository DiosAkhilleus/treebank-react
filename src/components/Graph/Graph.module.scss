.graph {
  background-color: rgb(242, 242, 242);
  margin-bottom: 0.2rem;

  path {
    stroke: #333;
    stroke-width : 0.5px;
    fill: none;
  }

  text {
    fill: #222;
    font-family: sans-serif;
  }

  :global {
    .nodes {
      fill: none;
    }

    .edgeLabels {
      text {
        font-size: 0.7rem;
      }
    }
  }
}

.node {
  :global {
    .label {
      div {
        font-size: 1rem;
        cursor: pointer;

        &:hover {
          background-color: #5BC8DC;
        }
      }
    }
  }
}

.active {
  :global {
    .label {
      div {
        background-color: rgb(246, 217, 24);
      }
    }
  }
}
