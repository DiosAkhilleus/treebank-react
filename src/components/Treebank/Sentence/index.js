import { SentenceConsumer } from './Sentence';

export default SentenceConsumer;
