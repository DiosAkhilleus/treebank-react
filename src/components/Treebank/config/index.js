import {
  getConfig,
  getColor,
} from './config';

export {
  getConfig,
  getColor,
};
