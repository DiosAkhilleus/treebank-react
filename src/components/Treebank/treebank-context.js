import { createContext } from 'react';

const TreebankContext = createContext();

export default TreebankContext;
