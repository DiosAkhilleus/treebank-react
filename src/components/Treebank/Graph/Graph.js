import React from 'react';

import TreebankContext from '../treebank-context';

import Graph from '../../Graph';

const GraphWithContext = () => (
  <TreebankContext.Consumer>
    {({
      sentence, active, setActive, config,
    }) => <Graph sentence={sentence} active={active} setActive={setActive} config={config} />}
  </TreebankContext.Consumer>
);

export default GraphWithContext;
