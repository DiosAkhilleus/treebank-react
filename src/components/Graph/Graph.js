import React from 'react';
import DagreGraph from 'dagre-d3-react';
import { curveBasis } from 'd3-shape';
import { object, func } from 'prop-types';

import { sentenceType, wordType } from '../../types';

import styles from './Graph.module.scss';

import { sentenceToGraph } from '../../lib/parsing';
import { getColor } from '../Treebank/config';

const dagreConfig = {
  rankdir: 'TB',
};

const nodeConfig = (config, active, { id, postag }) => {
  const color = getColor(config, postag);
  const isActive = active && active.$.id === id;
  const className = isActive ? [styles.node, styles.active].join(' ') : styles.node;

  return {
    labelStyle: `color: ${color}`,
    labelType: 'html',
    class: className,
  };
};

const configureNodes = (config, active, nodes) => (
  nodes.map((node) => {
    // eslint-disable-next-line no-param-reassign
    node.config = nodeConfig(config, active, node);

    return node;
  })
);

const linkConfig = {
  arrowheadStyle: 'display: none',
  curve: curveBasis,
};

const configureLinks = (links) => (
  links.map((link) => {
    // eslint-disable-next-line no-param-reassign
    link.config = linkConfig;

    return link;
  })
);

const Graph = ({
  sentence, active, toggleActive, config,
}) => {
  const { nodes, links } = sentenceToGraph(sentence);

  return (
    <DagreGraph
      nodes={configureNodes(config, active, nodes)}
      links={configureLinks(links)}
      zoomable
      className={styles.graph}
      config={dagreConfig}
      onNodeClick={({ original: { _word } }) => toggleActive(_word)}
    />
  );
};

Graph.propTypes = {
  sentence: sentenceType.isRequired,
  active: wordType,
  toggleActive: func.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  config: object.isRequired,
};

Graph.defaultProps = {
  active: null,
};

export default Graph;
