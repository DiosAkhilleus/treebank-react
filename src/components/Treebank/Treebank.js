import React, { useState } from 'react';
import { node, string } from 'prop-types';

import styles from './Treebank.module.scss';

import { xmlToJson } from '../../lib/parsing';
import { getConfig } from './config';

import TreebankContext from './treebank-context';

const sentenceFromJson = (treebankJson, chunk) => (
  treebankJson.treebank.sentence.find(({ $ }) => $.id && $.id === chunk)
);

const configFromJson = (treebankJson) => (
  getConfig(
    treebankJson.treebank.$['xml:lang'],
    treebankJson.treebank.$.format,
  )
);

const Treebank = ({ treebank, chunk, children }) => {
  const [active, setActive] = useState(null);
  const treebankJson = xmlToJson(treebank);
  const sentence = sentenceFromJson(treebankJson, chunk);
  const config = configFromJson(treebankJson);

  return (
    <TreebankContext.Provider value={{
      sentence,
      chunk,
      config,
      active,
      setActive,
    }}
    >
      <div className={styles.container}>
        {children}
      </div>
    </TreebankContext.Provider>
  );
};

Treebank.propTypes = {
  treebank: string.isRequired,
  chunk: string.isRequired,
  children: node,
};

Treebank.defaultProps = {
  children: null,
};

export default Treebank;
