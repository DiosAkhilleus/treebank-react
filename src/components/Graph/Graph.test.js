import React from 'react';
import renderer from 'react-test-renderer';

import { getConfig } from '../Treebank/config';

import Graph from './Graph';

it('renders a sentence', () => {
  const sentence = {
    $: {
      id: '1', document_id: 'test', subdoc: '2', span: '',
    },
    word: [
      {
        $: {
          id: '1', form: 'οὐκ', lemma: 'οὐ', postag: 'd--------', relation: 'AuxZ', head: '1',
        },
      },
      {
        $: {
          id: '2', form: 'ᾔδη', lemma: 'οἶδα', postag: 'v1slia---', relation: 'PRED', head: '0',
        },
      },
    ],
  };

  const component = (
    <Graph
      sentence={sentence}
      active={null}
      setActive={() => {}}
      config={getConfig('aldt', 'grc')}
    />
  );
  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});
