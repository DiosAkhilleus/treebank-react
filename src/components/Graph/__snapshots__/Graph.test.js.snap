// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a sentence 1`] = `
<svg
  className="graph"
  height="600"
  width="1000"
>
  <g />
</svg>
`;
